velox/functions/lib/SubscriptUtil.h

```diff
@@ -134,9 +124,11 @@ class SubscriptImpl : public exec::Subscript {
     auto indexArg = args[1];
 
     // Ensure map key type and second argument are the same.
-    VELOX_CHECK_EQ(mapArg->type()->childAt(0), indexArg->type());
+    VELOX_CHECK(mapArg->type()->childAt(0)->kindEquals(indexArg->type()));
```

Comment:
kindEquals is not enough because Varchar and Json both have the same type kind VARCHAR. You can use `mapArg->type()->equivalent(*indexArg->type())`.